docs/sql-data-sources-protobuf.md

```diff
@@ -0,0 +1,302 @@
+Dataset<Row> output = df
+.select(from_protobuf(col("value"), "org.spark.protobuf.User").as("user"))
+.where("user.favorite_color == \"red\"")
+.select(to_protobuf(col("user"), "org.spark.protobuf.User").as("value"));
+
+StreamingQuery query = output
+.writeStream()
+.format("kafka")
+.option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+.option("topic", "topic2")
+.start();
+
+{% endhighlight %}
+</div>
+<div data-lang="python" markdown="1">
+{% highlight python %}
+from pyspark.sql.protobuf.functions import from_protobuf, to_protobuf
+
+# `from_protobuf` and `to_protobuf` provides two schema choices. First, via the protobuf descriptor file, and then via the protobuf message class name.
+
+df = spark\
+.readStream\
+.format("kafka")\
+.option("kafka.bootstrap.servers", "host1:port1,host2:port2")\
+.option("subscribe", "topic1")\
+.load()
+
+# 1. Decode the Protobuf data into a struct;
+# 2. Filter by column `favorite_color`;
+# 3. Encode the column `user` in Protobuf format.
+# The Protobuf protoc command can be used to generate a protobuf descriptor file for give .proto file.
+output = df\
+.select(from_protobuf("value", protobufMessageName, descriptorFilePath).alias("user"))\
+.where('user.favorite_color == "red"')\
+.select(to_protobuf("user", protobufMessageName, descriptorFilePath).alias("value"))
+
+# Alternatively, you can decode and encode the SQL columns into protobuf format using protobuf class name.
+# The specified Protobuf class must match the data, otherwise the behavior is undefined: it may fail or return arbitrary result. The jar containing Java class should be shaded. Specifically, `com.google.protobuf.*` should be shaded to `org.sparkproject.spark-protobuf.protobuf.*`.
+output = df\
+.select(from_protobuf("value", "org.spark.protobuf.User").alias("user"))\
+.where('user.favorite_color == "red"')\
+.select(to_protobuf("user", "org.spark.protobuf.User").alias("value"))
+
+query = output\
+.writeStream\
+.format("kafka")\
+.option("kafka.bootstrap.servers", "host1:port1,host2:port2")\
+.option("topic", "topic2")\
+.start()
+
+{% endhighlight %}
+</div>
+</div>
+
+## Supported types for Protobuf -> Spark SQL conversion
+Currently Spark supports reading [protobuf scalar types](https://developers.google.com/protocol-buffers/docs/proto3#scalar), [enum types](https://developers.google.com/protocol-buffers/docs/proto3#enum), [nested type](https://developers.google.com/protocol-buffers/docs/proto3#nested), and [maps type](https://developers.google.com/protocol-buffers/docs/proto3#maps) under messages of Protobuf.
+One common issue that can arise when working with Protobuf data is the presence of circular references. In Protobuf, a circular reference occurs when a field refers back to itself or to another field that refers back to the original field. This can cause issues when parsing the data, as it can result in infinite loops or other unexpected behavior.
```

Comment:
I think we can move recursive protobuf to a sub section. Right now this directly jumps into this. 
This is not a new feature. You implemented in the first version itself :).

Also good to include an example. We can even move this to further down.

Reply:
I have a suggestion for rephrasing this section, but you can take attempt at it. An example would clarify things a lot for users.

Reply:
@rangadi, I have made the suggested changes. Could you please review them again?